[Bug]: Wrong type `policy` argument of `BaseAlgorithm` and `OffPolicyAlgorithm`

### 🐛 Bug

`policy` argument of `BaseAlgorithm` and `OffPolicyAlgorithm` has the type `Type[BasePolicy]` when it should be `Union[str, Type[BasePolicy]]`.

https://github.com/DLR-RM/stable-baselines3/blob/b77a0667b236cf52d71bf6f0b98b1cfa5ebe3ab6/stable_baselines3/common/off_policy_algorithm.py#L78

https://github.com/DLR-RM/stable-baselines3/blob/b77a0667b236cf52d71bf6f0b98b1cfa5ebe3ab6/stable_baselines3/common/base_class.py#L92

### To Reproduce

```shell
mypy stable_baselines3
```


### Relevant log output / Error message

```shell
stable_baselines3/dqn/dqn.py:99: error: Argument 1 to "__init__" of "OffPolicyAlgorithm" has incompatible type "Union[str, Type[DQNPolicy]]"; expected "Type[BasePolicy]"
```


### System Info

OS: macOS-12.6-x86_64-i386-64bit Darwin Kernel Version 21.6.0: Mon Aug 22 20:20:05 PDT 2022; root:xnu-8020.140.49~2/RELEASE_ARM64_T8101
Python: 3.9.13
Stable-Baselines3: 1.7.0a0
PyTorch: 1.12.1
GPU Enabled: False
Numpy: 1.23.3
Gym: 0.21.0

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] I have provided a minimal working example to reproduce the bug
- [X] I've used the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

--- a/docs/misc/changelog.rst
+++ b/docs/misc/changelog.rst
@@ -23,6 +23,7 @@ SB3-Contrib
 Bug Fixes:
 ^^^^^^^^^^
 - Fix return type of ``evaluate_actions`` in ``ActorCritcPolicy`` to reflect that entropy is an optional tensor (@Rocamonde)
+- Fix type annotation of ``policy`` in ``BaseAlgorithm`` and ``OffPolicyAlgorithm``
 
 Deprecations:
 ^^^^^^^^^^^^^

--- a/stable_baselines3/common/base_class.py
+++ b/stable_baselines3/common/base_class.py
@@ -60,7 +60,7 @@ class BaseAlgorithm(ABC):
     """
     The base of RL algorithms
 
-    :param policy: Policy object
+    :param policy: The policy model to use (MlpPolicy, CnnPolicy, ...)
     :param env: The environment to learn from
                 (if registered in Gym, can be str. Can be None for loading trained models)
     :param learning_rate: learning rate for the optimizer,
@@ -89,7 +89,7 @@ class BaseAlgorithm(ABC):
 
     def __init__(
         self,
-        policy: Type[BasePolicy],
+        policy: Union[str, Type[BasePolicy]],
         env: Union[GymEnv, str, None],
         learning_rate: Union[float, Schedule],
         policy_kwargs: Optional[Dict[str, Any]] = None,

--- a/stable_baselines3/common/off_policy_algorithm.py
+++ b/stable_baselines3/common/off_policy_algorithm.py
@@ -28,7 +28,7 @@ class OffPolicyAlgorithm(BaseAlgorithm):
     """
     The base for Off-Policy algorithms (ex: SAC/TD3)
 
-    :param policy: Policy object
+    :param policy: The policy model to use (MlpPolicy, CnnPolicy, ...)
     :param env: The environment to learn from
                 (if registered in Gym, can be str. Can be None for loading trained models)
     :param learning_rate: learning rate for the optimizer,
@@ -75,7 +75,7 @@ class OffPolicyAlgorithm(BaseAlgorithm):
 
     def __init__(
         self,
-        policy: Type[BasePolicy],
+        policy: Union[str, Type[BasePolicy]],
         env: Union[GymEnv, str],
         learning_rate: Union[float, Schedule],
         buffer_size: int = 1_000_000,  # 1e6

